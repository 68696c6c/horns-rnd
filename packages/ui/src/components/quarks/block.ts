import {
  chromatic,
  Chromatic,
  contained,
  Contained,
  padded,
  Padded,
} from '../../traits'

export interface BlockProps extends Chromatic, Contained, Padded {}

export const blockStyles = [chromatic, contained, padded]
