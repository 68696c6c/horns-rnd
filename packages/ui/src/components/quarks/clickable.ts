// Clickables are interactive elements that can be clicked on, like buttons, links, and nav items.
import styled from '@emotion/styled'
import { css } from '@emotion/react'

import { Cursor, Font, HoverState, StatusState } from '../../config'
import {
  Bordered,
  bordered,
  Chromatic,
  chromatic,
  chromaticText,
  Parent,
  interactive,
  Interactive,
  Padded,
  padded,
  Rounded,
  rounded,
  Styled,
  Typographic,
  typographic,
} from '../../traits'

export enum LinkVariant {
  Link = 'link',
  Button = 'button',
}

export interface ButtonProps
  extends Parent,
    Bordered,
    Chromatic,
    Interactive,
    Padded,
    Rounded,
    Typographic {}

const buttonStyles = ({
  theme,
  border,
  color,
  cursor,
  font,
  padding,
  radius,
}: Styled & ButtonProps) => {
  const { buttons } = theme
  return [
    css`
      display: inline-block;
    `,
    chromatic,
    bordered({ theme, border, borderDefault: buttons.border }),
    interactive({
      theme,
      cursor,
      cursorDefault: Cursor.Pointer,
      hoverStyles: [chromatic({ theme, color, state: HoverState.Hover })],
      activeStyles: [chromatic({ theme, color, state: HoverState.Active })],
      inactiveStyles: [
        chromatic({ theme, color, state: StatusState.Inactive }),
      ],
    }),
    padded({ theme, padding, paddingDefault: buttons.padding }),
    rounded({ theme, radius, radiusDefault: buttons.radius }),
    typographic({ theme, font, fontDefault: Font.Button }),
  ]
}

export const StyledLinkButton = styled.a(buttonStyles)

export const StyledButton = styled.button(buttonStyles)

export interface LinkProps extends ButtonProps {
  variant?: LinkVariant
}

export const linkStyles = ({
  theme,
  color,
  cursor,
  font,
}: Styled & LinkProps) => {
  const chromaticArgs = { theme, color }
  const typographicArgs = { theme, font, fontDefault: Font.Link }
  return [
    chromaticText,
    typographic({ ...typographicArgs }),
    interactive({
      theme,
      cursor,
      cursorDefault: Cursor.Pointer,
      hoverStyles: [
        chromaticText({ ...chromaticArgs, state: HoverState.Hover }),
        typographic({ ...typographicArgs, state: HoverState.Hover }),
      ],
      activeStyles: [
        chromaticText({ ...chromaticArgs, state: HoverState.Active }),
        typographic({ ...typographicArgs, state: HoverState.Active }),
      ],
      inactiveStyles: [
        chromaticText({ ...chromaticArgs, state: StatusState.Inactive }),
        typographic({ ...typographicArgs, state: StatusState.Inactive }),
      ],
    }),
    () => css`
      &:visited {
        ${chromaticText({ ...chromaticArgs, state: HoverState.Hover })}
        ${typographic({ ...typographicArgs, state: HoverState.Hover })}
      }
    `,
  ]
}

export const StyledLink = styled.a(linkStyles)

export const styleLink = (CustomComponent: any) =>
  styled(CustomComponent)(linkStyles)

export const styleButton = (CustomComponent: any) =>
  styled(CustomComponent)(buttonStyles)
