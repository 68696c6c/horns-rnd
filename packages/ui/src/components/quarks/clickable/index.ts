import styled from '@emotion/styled'

import {
  Styled,
  Component,
  Bordered,
  bordered,
  Chromatic,
  chromatic,
  chromaticText,
  Padded,
  padded,
} from '../../../traits'

export interface ButtonProps
  extends Styled,
    Component,
    Bordered,
    Chromatic,
    Padded {}

export enum LinkVariant {
  Link = 'link',
  Button = 'button',
}

export interface BaseLinkProps extends ButtonProps {
  variant?: LinkVariant
}

export const Link = styled.a<BaseLinkProps>(
  ({ theme, color }: BaseLinkProps) => [chromaticText(theme, color)],
)

export const LinkButton = styled.a<ButtonProps>(
  ({ theme, border, color, padding }: ButtonProps) => [
    chromatic(theme, color),
    bordered(theme, border, theme.buttons.border),
    padded(theme, padding, theme.buttons.padding),
  ],
)

// type LinkVariantProps = {
//   Tag:
// }
//
// export const getLinkVariantProps = (variant: string): LinkVariantProps => {
//   if (variant === 'button') {
//     return {
//       Tag: LinkButton,
//     }
//   }
//   return {
//     Tag: Link,
//   }
// }
