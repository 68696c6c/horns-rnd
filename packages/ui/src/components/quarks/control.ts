// Controls are form elements.
import { EventHandler } from 'react'
import { css } from '@emotion/react'

import {
  ControlOption,
  Cursor,
  Font,
  HoverState,
  StatusState,
} from '@horns/theme'

import {
  Bordered,
  bordered,
  chromaticControl,
  ChromaticNotification,
  interactive,
  Interactive,
  Padded,
  padded,
  Rounded,
  rounded,
  Styled,
  Typographic,
  typographic,
} from '../../traits'

export interface ControlOptionProps {
  options?: ControlOption[]
}

export interface ControlProps
  extends Bordered,
    ChromaticNotification,
    Interactive,
    Padded,
    Rounded,
    Typographic {
  id?: string
  name?: string
  value?: string | number
  defaultValue?: string | number
  onKeyUp?: EventHandler<any>
  onChange?: EventHandler<any>
  required?: boolean
  multiline?: boolean
  multiple?: boolean
  autoComplete?: string
}

export const controlStyles = ({
  theme,
  border,
  color,
  cursor,
  padding,
  radius,
}: Styled & ControlProps) => {
  const { controls } = theme
  return [
    chromaticControl,
    bordered({ theme, border, borderDefault: controls.border }),
    interactive({
      theme,
      cursor,
      cursorDefault: Cursor.Text,
      hoverStyles: [
        chromaticControl({ theme, color, state: HoverState.Hover }),
      ],
      activeStyles: [
        chromaticControl({ theme, color, state: HoverState.Active }),
      ],
      inactiveStyles: [
        chromaticControl({ theme, color, state: StatusState.Inactive }),
      ],
    }),
    padded({ theme, padding, paddingDefault: controls.padding }),
    rounded({ theme, radius, radiusDefault: controls.radius }),
    typographic({ theme, font: Font.Control }),
    ({ multiple, multiline }: ControlProps) => {
      // Force the element height to match the line-height to ensure that inputs that have controls
      // inside them (e.g. input type="datetime-local") don't end up a different size than standard inputs.
      const f = theme.typography.control.base
      if (multiple) {
        return css`
          min-height: ${f.letting};
        `
      }
      return multiline
        ? css``
        : css`
            height: ${f.letting};
            min-width: ${f.letting};
          `
    },
    () => css`
      display: inline-flex;
      box-sizing: content-box;
      vertical-align: middle;
    `,
  ]
}
