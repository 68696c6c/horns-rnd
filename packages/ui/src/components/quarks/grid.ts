import {
  Gridded,
  gridded,
  padded,
  Padded,
  Parent,
  responsive,
  Responsive,
  Styled,
} from '../../traits'

export interface GridProps extends Parent, Gridded, Padded, Responsive {}

export const gridStyles = ({
  theme,
  breakpoint,
  gapped,
  gap,
}: Styled & GridProps) => [
  padded,
  responsive({
    theme,
    breakpoint,
    responsiveStyles: gridded({ theme, gapped, gap }),
  }),
]
