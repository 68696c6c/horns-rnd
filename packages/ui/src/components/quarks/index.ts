// Quarks are styled components like atoms, but are for internal use only and are not exported
// outside of the components module.  They are used as a way for otherwise unrelated atoms, like
// Button and Link, to share certain styles with each other.  Like atoms, they are composed of
// traits and export styled components, making them distinctly different than traits, which are
// neither composed nor components and exist only to apply theming to specific css properties.
export { BlockProps, blockStyles } from './block'
export {
  StyledButton,
  LinkVariant,
  ButtonProps,
  Anchor,
  BaseLinkProps,
  LinkProps,
  StyledLinkButton,
  StyledLink,
  styleLink,
  styleButton,
  BaseNavItemProps,
  NavItemLayout,
  NavItemVariant,
  getLinkVariantTag,
  styleCustomLinkTag,
} from './clickable'
export { ControlProps, controlStyles } from './control'
export { GridProps, gridStyles } from './grid'
export {
  MenuProps,
  NavMenuProps,
  menuStyles,
  navMenuStyles,
  MenuContainer,
} from './menu'
export { MessageProps, messageStyles } from './message'
export {
  HTMLDataset,
  TableProps,
  TableRows,
  TableRow,
  TableData,
  tableStyles,
  getTableData,
  OverflowProps,
  Overflow,
} from './table'
