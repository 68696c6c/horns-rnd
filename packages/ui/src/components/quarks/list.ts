import { css } from '@emotion/react'

import { Colorway } from '@horns/theme'

import styled from '@emotion/styled'
import {
  chromaticText,
  Chromatic,
  gridded,
  Gridded,
  Parent,
  Styled,
  typographic,
  Typographic,
} from '../../traits'

export interface ListMarkerProps {
  markerColor?: Colorway
}

export interface LIProps
  extends Chromatic,
    Typographic,
    Parent,
    ListMarkerProps {}

export const LI = styled.li<LIProps>(chromaticText, typographic)

export interface ListProps
  extends Chromatic,
    Gridded,
    Typographic,
    Parent,
    ListMarkerProps {}

export const listStyles = ({
  theme,
  color,
  markerColor,
}: Styled & ListProps) => [
  chromaticText,
  gridded,
  typographic,
  () => {
    const marker = theme.colors[markerColor || color || Colorway.Typography]
    // const c = getItemColors(theme, color, markerColor, false)
    // TODO: spacing
    // const spacing = theme.sizing.getPX(listDefaults.spacingLeft)
    // // padding-left: ${spacing};
    // Note that in browsers that do not support the ::marker selector (e.g. Chrome), combining it
    // with any other supported selectors will cause the entire rule to fail, hence the separate
    // blocks for ::marker and .list-item-marker.
    return css`
      margin: 0;
      grid-template-rows: auto;
      padding-left: 1em;
      li::marker {
        color: ${marker.base.base};
      }
      .marker {
        color: ${marker.base.base};
      }
    `
  },
]
