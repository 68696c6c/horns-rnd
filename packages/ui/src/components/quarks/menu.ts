import styled from '@emotion/styled'
import { css } from '@emotion/react'

import { chromatic, Chromatic, shadowed, Shadowed, Styled } from '../../traits'

export interface MenuProps extends Shadowed {
  open?: boolean
}

export const menuStyles = () => [
  ({ theme, open, shadow }: Styled & MenuProps) => {
    if (open) {
      return shadowed({ theme, shadow })
    }
    return css`
      display: none;
    `
  },
  () => css`
    position: absolute;
    box-sizing: border-box;
    min-width: 100%;
  `,
]

export interface NavMenuProps extends MenuProps, Chromatic {}

export const navMenuStyles = () => [chromatic, menuStyles]

export const MenuContainer = styled.span<MenuProps>(
  () =>
    css`
      position: relative;
    `,
  ({ theme, open, shadow }) =>
    open &&
    css`
    > * {
      ${shadowed({ theme, shadow })}
    `,
)
