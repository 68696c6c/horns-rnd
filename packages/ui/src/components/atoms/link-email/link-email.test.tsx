import React from 'react'

import {
  assertButtonStateStyles,
  assertLinkStateStyles,
  render,
  screen,
} from '../../test'
import {
  BorderStyle,
  Colorway,
  Font,
  HeadingLevel,
  Size,
} from '../../../config'
import { LinkVariant } from '../../quarks'

import { LinkEmail } from '.'

const props = {
  email: 'example@example.com',
  subject: 'example subject',
  body: 'example body',
}

describe('LinkEmail', () => {
  it('should render as default', () => {
    const { container } = render(<LinkEmail {...props}>example</LinkEmail>)
    expect(container).toMatchSnapshot()
  })

  it('should render as a button', () => {
    const { container } = render(
      <LinkEmail {...props} variant={LinkVariant.Button}>
        example
      </LinkEmail>,
    )
    expect(container).toMatchSnapshot()
  })

  describe('link variant', () => {
    it('should render as default', () => {
      const { container } = render(
        <LinkEmail {...props} variant={LinkVariant.Link}>
          example
        </LinkEmail>,
      )
      expect(container).toMatchSnapshot()
    })
    it.each(Object.values(Colorway))('should render colorway %s', (color) => {
      const { container } = render(
        <LinkEmail {...props} variant={LinkVariant.Link} color={color}>
          {color}
        </LinkEmail>,
      )
      expect(container).toMatchSnapshot()
    })
    it.each([...Object.values(Font), ...Object.values(HeadingLevel)])(
      'should render font %s',
      (font) => {
        const { container } = render(
          <LinkEmail {...props} variant={LinkVariant.Link} font={font}>
            {font}
          </LinkEmail>,
        )
        expect(container).toMatchSnapshot()
      },
    )
    describe('ui states', () => {
      it('should render default state', () => {
        const text = 'link-default'
        render(
          <LinkEmail {...props} variant={LinkVariant.Link}>
            {text}
          </LinkEmail>,
        )
        assertLinkStateStyles(screen.getByText(text), '', {
          color: 'rgb(1, 1, 1)',
          'text-decoration-color': 'rgb(1, 1, 1)',
          'text-decoration-line': 'underline',
          'text-decoration-style': 'solid',
          'font-style': 'normal',
        })
      })
      it('should render hover state', () => {
        const text = 'link-hover'
        render(
          <LinkEmail {...props} variant={LinkVariant.Link}>
            {text}
          </LinkEmail>,
        )
        assertLinkStateStyles(screen.getByText(text), ':hover', {
          color: 'rgb(1, 1, 1)',
          'text-decoration-color': 'rgb(1, 1, 1)',
          'text-decoration-line': 'underline',
          'text-decoration-style': 'solid',
          'font-style': 'italic',
        })
      })
      it('should render active state', () => {
        const text = 'link-active'
        render(
          <LinkEmail {...props} variant={LinkVariant.Link}>
            {text}
          </LinkEmail>,
        )
        assertLinkStateStyles(screen.getByText(text), ':active', {
          color: 'rgb(1, 1, 1)',
          'text-decoration-color': 'rgb(1, 1, 1)',
          'text-decoration-line': 'underline',
          'text-decoration-style': 'double',
          'font-style': 'italic',
        })
      })
      it('should render visited state', () => {
        const text = 'link-visited'
        render(
          <LinkEmail {...props} variant={LinkVariant.Link}>
            {text}
          </LinkEmail>,
        )
        assertLinkStateStyles(screen.getByText(text), ':visited', {
          color: 'rgb(1, 1, 1)',
          'text-decoration-color': 'rgb(1, 1, 1)',
          'text-decoration-line': 'underline',
          'text-decoration-style': 'solid',
          'font-style': 'italic',
        })
      })
    })
  })

  describe('button variant', () => {
    it('should render as default', () => {
      const { container } = render(
        <LinkEmail {...props} variant={LinkVariant.Button}>
          example
        </LinkEmail>,
      )
      expect(container).toMatchSnapshot()
    })
    it.each(Object.values(Colorway))('should render colorway %s', (color) => {
      const { container } = render(
        <LinkEmail {...props} variant={LinkVariant.Button} color={color}>
          {color}
        </LinkEmail>,
      )
      expect(container).toMatchSnapshot()
    })
    it.each(Object.values(BorderStyle))(
      'should render border style %s',
      (style) => {
        const { container } = render(
          <LinkEmail
            {...props}
            variant={LinkVariant.Button}
            border={{ all: { style, width: Size.Small } }}
          >
            {style}
          </LinkEmail>,
        )
        expect(container).toMatchSnapshot()
      },
    )
    it.each(Object.values(Size))('should render border width %s', (size) => {
      const { container } = render(
        <LinkEmail
          {...props}
          variant={LinkVariant.Button}
          border={{ all: { style: BorderStyle.Solid, width: size } }}
        >
          {size}
        </LinkEmail>,
      )
      expect(container).toMatchSnapshot()
    })
    it.each(Object.values(Size))('should render padding %s', (size) => {
      const { container } = render(
        <LinkEmail {...props} variant={LinkVariant.Button} padding={size}>
          {size}
        </LinkEmail>,
      )
      expect(container).toMatchSnapshot()
    })
    it.each(Object.values(Size))('should render radius %s', (size) => {
      const { container } = render(
        <LinkEmail {...props} variant={LinkVariant.Button} radius={size}>
          {size}
        </LinkEmail>,
      )
      expect(container).toMatchSnapshot()
    })
    it.each([...Object.values(Font), ...Object.values(HeadingLevel)])(
      'should render font %s',
      (font) => {
        const { container } = render(
          <LinkEmail {...props} variant={LinkVariant.Button} font={font}>
            {font}
          </LinkEmail>,
        )
        expect(container).toMatchSnapshot()
      },
    )
    describe('ui states', () => {
      it('should render default state', () => {
        const text = 'link-button-default'
        render(
          <LinkEmail {...props} variant={LinkVariant.Button}>
            {text}
          </LinkEmail>,
        )
        assertButtonStateStyles(screen.getByText(text), '', {
          'background-color': 'rgb(255, 255, 255)',
        })
      })
      it('should render hover state', () => {
        const text = 'link-button-hover'
        render(
          <LinkEmail {...props} variant={LinkVariant.Button}>
            {text}
          </LinkEmail>,
        )
        assertButtonStateStyles(screen.getByText(text), ':hover', {
          'background-color': 'rgb(247, 247, 247)',
        })
      })
      it('should render active state', () => {
        const text = 'link-button-active'
        render(
          <LinkEmail {...props} variant={LinkVariant.Button}>
            {text}
          </LinkEmail>,
        )
        assertButtonStateStyles(screen.getByText(text), ':active', {
          'background-color': 'rgb(240, 240, 240)',
        })
      })
    })
  })
})
