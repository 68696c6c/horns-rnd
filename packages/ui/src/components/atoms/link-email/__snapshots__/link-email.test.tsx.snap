// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkEmail should render as a button 1`] = `
.emotion-0 {
  border-color: rgb(247, 247, 247);
  background-color: rgb(255, 255, 255);
  color: rgb(1, 1, 1);
  border-top-width: 2px;
  border-top-style: solid;
  border-right-width: 2px;
  border-right-style: solid;
  border-bottom-width: 2px;
  border-bottom-style: solid;
  border-left-width: 2px;
  border-left-style: solid;
  cursor: pointer;
  padding-top: 12px;
  padding-right: 16px;
  padding-bottom: 12px;
  padding-left: 16px;
  font-family: Helvetica,sans-serif;
  font-kerning: normal;
  font-size: medium;
  font-style: normal;
  font-weight: 700;
  text-align: left;
  text-decoration-line: none;
  text-transform: none;
}

.emotion-0:disabled,
.emotion-0.inactive {
  border-color: rgb(165, 165, 165);
  background-color: rgb(191, 191, 191);
  color: rgb(1, 1, 1);
}

.emotion-0:not(:disabled):not(.inactive):hover {
  border-color: rgb(240, 240, 240);
  background-color: rgb(247, 247, 247);
  color: rgb(1, 1, 1);
}

.emotion-0:not(:disabled):not(.inactive):active {
  border-color: rgb(232, 232, 232);
  background-color: rgb(240, 240, 240);
  color: rgb(1, 1, 1);
}

<div>
  <a
    class="emotion-0"
    href="mailto:example@example.com?subject=example subject&body=example body"
  >
    example
  </a>
</div>
`;

exports[`LinkEmail should render as default 1`] = `
.emotion-0 {
  color: rgb(1, 1, 1);
  font-family: Helvetica,sans-serif;
  font-kerning: normal;
  font-size: medium;
  font-style: normal;
  font-weight: 400;
  text-align: left;
  text-decoration-line: none;
  text-transform: none;
  cursor: pointer;
}

.emotion-0:disabled,
.emotion-0.inactive {
  color: rgb(1, 1, 1);
  font-family: Helvetica,sans-serif;
  font-kerning: normal;
  font-size: medium;
  font-style: normal;
  font-weight: 400;
  text-align: left;
  text-decoration-line: underline;
  text-decoration-style: double;
  text-transform: none;
}

.emotion-0:not(:disabled):not(.inactive):hover {
  color: rgb(1, 1, 1);
  font-family: Helvetica,sans-serif;
  font-kerning: normal;
  font-size: medium;
  font-style: normal;
  font-weight: 400;
  text-align: left;
  text-decoration-line: underline;
  text-decoration-style: double;
  text-transform: none;
}

.emotion-0:not(:disabled):not(.inactive):active {
  color: rgb(1, 1, 1);
  font-family: Helvetica,sans-serif;
  font-kerning: normal;
  font-size: medium;
  font-style: normal;
  font-weight: 400;
  text-align: left;
  text-decoration-line: underline;
  text-decoration-style: double;
  text-transform: none;
}

<div>
  <a
    class="emotion-0"
    href="mailto:example@example.com?subject=example subject&body=example body"
  >
    example
  </a>
</div>
`;
